packages/run-protocol/src/contractSupport.js

```diff
@@ -48,6 +48,12 @@ export const assertOnlyKeys = (proposal, keys) => {
  * the gain and a loss on the `fromSeat`. The gain/loss are typically from the
  * give/want respectively of a proposal. The `key` is the allocation keyword.
  *
+ * Warning! if the amounts are both empty, this won't add a staged allocation to
+ * either seat, so the seats must only be passed to reallocate() if something
+ * else does stage an allocation. It would be wrong to throw here, since the
+ * common pattern of use is to call this twice, with one of the two calls not
```

Comment:
When I see a warning in a comment, I assume someone will miss it and I try to think how we could enforce it programmatically.

Searching for all uses, there are basically two instances of the same logic that happens in vault and runStake:
```
    stageDelta(clientSeat, vaultSeat, giveColl, wantColl, 'Collateral');
    // `wantRUN` is allocated in the reallocate and mint operation, and so not here
    stageDelta(clientSeat, vaultSeat, giveRUN, helper.emptyDebt(), 'RUN');
```
```
    stageDelta(clientSeat, vaultSeat, giveColl, wantColl, KW.Attestation);
    stageDelta(clientSeat, vaultSeat, giveRUN, emptyDebt, KW.Debt);
```

So I content that _all_ uses call two times and because the function is so small we don't need to support the single call case. In which case the function can check itself whether any allocation was staged.

A clean way to do that would be to pass an array of deltas, which would have the added benefit of working for the n=1 case and n>2 cases. I.e. the function signature would be instead,
```
 * @param {ZCFSeat} fromSeat
 * @param {ZCFSeat} toSeat
 * @param {Array<[losses: Amount, gains: Amount, key: Keyword>} fromDeltas
```

Reply:
I went through the same analysis (summarized in the top comment of this PR).

I don't think that catching the problem in stageDelta or its revision improves the situation. The reallocation will fail after an incorrect call, and the transaction will be aborted. The problem I tried to address is that the error message is inscrutable. In vaultFactory, we can detect it before calling stageDelta, and give a more understandable response. After analysis there didn't seem to be a need to do the same in runStake, since it already has guards that prevent requests for attestations for empty amounts.

Reply:
> Wouldn't throwing when no allocations are staged be scrutable?

somewhat. The user has no concept of allocations or reallocations, or of what seats are involved internally so rephrasing in their terms seems more helpful. 

If this seems unnecessary, I'm alright with that decision. I feel like we've established that there aren't routes through this code that lead to problematic states.

Reply:
I removed the warning.

If you call `adjustBalances` in `runStake` with empty changes, it transfers empty attestations, so the reallocate is happy.

We only need the check for zero in vaults.